### ISSUE TITLE
How to judge a stranger in my dataset?

### ISSUE THREAD

**Issue by Tomhouxin** [davidsandberg/facenet#622]
.  I want to achieve real_time_face_recognition py, it mistook a stranger people in the dataset, it should be the most similar as the stranger in the dataset, then how should I process to determine the stranger unknowner?
  Thank you very much

**Comment by Shahnawazgrewal** [comment]
Can you please rephrase your question. @Tomhouxin

**Comment by Tomhouxin** [comment]
When I was doing face recognition, I found that I could only recognize faces in data sets. The problem is when I identify a face that is not in the data set. It actually discriminates me into the most similar human face in a dataset, not a stranger. @Shahnawazgrewal

**Comment by Shahnawazgrewal** [comment]
for previously unseen identities, you need to extract embedding and then a train classifier to recognize. @Tomhouxin

**Comment by Tomhouxin** [comment]
According to my understanding, I just need to decide that people I don't know are strangers, and I don't need to put them into the classifier. @Shahnawazgrewal

**Comment by jax79sg** [comment]
Maybe find out the distribution of confidence levels of different classes for unknown faces. Then from there, see if can come up with a heuristic to determine if its unknown.

**Comment by Tomhouxin** [comment]
You have a good idea, but people you know and people you don't know are very similar in terms of SVM classification results.
@jax79sg

**Comment by RedBugYO** [comment]
@Tomhouxin How about create a category called as "unknown" and put a lots of face image which you don't have to classify in it then re-train the classifier. I think it may make the classifier better for usage.

**Comment by Tomhouxin** [comment]
This problem is solved by setting a threshold for the results of the classifier. At the very beginning, I saw that the results of the classification were very close, and there was no threshold. Now when I train the classifier, I add about 100 faces to each person. The result of the classification is obviously different, and then you can set the threshold.（这个问题解决了，把分类器出来的结果设一个阈值。最开始我看分类的结果都很接近，无法设阈值。现在我训练分类器的时候每人加到100张人脸左右。分类出来的结果就很明显有差距了，这时候就可以设阈值） @RedBugYO

**Comment by niu-niu** [comment]
@Tomhouxin try other classifier, if you find the classification results are very close to each other.

**Comment by jiangzidong** [comment]
hi @Tomhouxin What does the "threshold" mean in the SVM classifier？
1） check the probability by `model.predict_proba(emb)`, if the max one is below a threshold, then it is a stranger?
or
2) add a new class called "unknown" in to the classifier?

**Comment by DawnHH** [comment]
@Tomhouxin 所以你还是增加了一个新类叫stranger吗？和embedding预测的类别相似度小于阈值的都被分类为这个类别里？

**Comment by rain2008204** [comment]
@DawnHH only add a threshold

**Comment by Tomhouxin** [comment]
@DawnHH yes

**Comment by menon92** [comment]
Hello @Tomhouxin , 
Did you solve your problem ?

**Comment by Tomhouxin** [comment]
@menon92 , yes !

**Comment by menon92** [comment]
Thanks @Tomhouxin
Could you please help me that how can I do this. I have the same problem. I try to calculate `l2 distance' of two image. But it's not working well

Thanks

**Comment by Tomhouxin** [comment]
sorry，I'm not sure how good it is to calculate the distance. I'm using the classifier (SVM), but I'm also going to try to calculate the distance directly

**Comment by menon92** [comment]
@Tomhouxin 
Do you just use classifier (SVM) prediction with a fix threshold to  judge a stranger ?

**Comment by Tomhouxin** [comment]
@menon92 ,yes ,you are right

**Comment by menon92** [comment]
@Tomhouxin 
But it's not working for me. what could be the cause of it . However my image quality is not very good

**Comment by Tomhouxin** [comment]
I don't know exactly what your question is, can't you find the appropriate threshold?

**Comment by menon92** [comment]
@Tomhouxin 
 Yes, I can't find the appropriate threshold for classifying the unknown images.

**Comment by Tomhouxin** [comment]
Well, please read the comments carefully. They say how to solve it.

**Comment by ZhangYuef** [comment]
> hi @Tomhouxin What does the "threshold" mean in the SVM classifier？
> 1） check the probability by `model.predict_proba(emb)`, if the max one is below a threshold, then it is a stranger?
> or
> 2) add a new class called "unknown" in to the classifier?

I think their method is as the same as the first one you mentioned.
